### Data Wrangling

In this notebook, will go through the data wrangling process for our project. The sponser, WebFilms, has provided two sources of data for us to use:

* What they currently have in their database for movie ratings (Recommendation System.csv)
* The Movie ID and Title keys (Movie_Id_Titles).

The sponser is clear that they don't want us using any other data than the data provided so we will work with the data that is provided.

In [1]:
# First, let's import relevant packages
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# Second, let's first take a look at the data provided regarding ratings
#import the csv into a Dataframe
ratings_df = pd.read_csv('Recommendation System.csv')

In [3]:
#Let's look at the DataFrame
ratings_df.head()

0   50  5  881250949
0    0  172  5  881250949
1    0  133  1  881250949
2  196  242  3  881250949
3  186  302  3  891717742
4   22  377  1  878887116

According to the documentation, we have the following categories of information:
* user_id: The unique identifier for a user
* item_id: The unique identifier for a movie
* rating: the rating that a specific user gave a specific movie on a scale of 1 to 5
* timestamp: when in time the rating was given

This means we'll need to apply column labels to our dataframe

In [4]:
#First, let's create a list of our column labels
column_names = ['user_id', 'movie_id', 'rating', 'timestamp']

#Second, let's apply our column names to the dataframe
ratings_df.columns = column_names

#Check to see that everything worked
ratings_df.head()

user_id  movie_id  rating  timestamp
0        0       172       5  881250949
1        0       133       1  881250949
2      196       242       3  881250949
3      186       302       3  891717742
4       22       377       1  878887116

The first thing we should consider is whether there are any empty elements in our table or anything that may make it difficult to conduct exploratory data analysis (EDA). Therefore, let's explore our data a little to get a sense of what we have.

In [5]:
#Let's get some summary information about our dataframe
ratings_df.describe()

user_id       movie_id         rating     timestamp
count  100002.000000  100002.000000  100002.000000  1.000020e+05
mean      462.475500     425.524670       3.529849  8.835288e+08
std       266.619776     330.797313       1.125700  5.343812e+06
min         0.000000       1.000000       1.000000  8.747247e+08
25%       254.000000     175.000000       3.000000  8.794487e+08
50%       447.000000     322.000000       4.000000  8.828269e+08
75%       682.000000     631.000000       4.000000  8.882600e+08
max       943.000000    1682.000000       5.000000  8.932866e+08

In [6]:
ratings_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100002 entries, 0 to 100001
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype
---  ------     --------------   -----
 0   user_id    100002 non-null  int64
 1   movie_id   100002 non-null  int64
 2   rating     100002 non-null  int64
 3   timestamp  100002 non-null  int64
dtypes: int64(4)
memory usage: 3.1 MB


There doesn't appear to be any evidence that we are missing information or any indication that we should be concerned at a cursory level about the data. That said, let's calculate the number of unique entries for user_id and movie_id to get a sense of how many unique elements are in each column

In [7]:
#Count number of unique users
print(f'Number of unique users: {len(ratings_df.user_id.unique())}')

Number of unique users: 944


In [8]:
#Count number of unique movies
print(f'Number of unique movies: {len(ratings_df.movie_id.unique())}')

Number of unique movies: 1682


This would imply that we have 944 Users and 1682 Movies in our database, which lines up with the information provided by the sponsor. Therefore we should be confident that the data is accurate for these two columns. The next step is to check the ratings for the rows to make sure they are integers between 1-5. Although the descriptive statistics would imply they are, can't hurt to be a little extra careful.

In [9]:
#Get all the unique elements in the rating column
print(ratings_df.rating.unique())

[5 1 3 2 4]


This gives us confidence that our data for ratings is proper. One last bit of information is that the sponsor provided the movie title associated with the movie_ids in a different table. While it's not clear this will be helpful, we can add them to the dataframe for now and always drop it if necesssary.

In [10]:
#Load table with movie_id and title
movie_table = pd.read_table('Movie_Id_Titles', delimiter=',', header = 0)

In [11]:
#Let's check to make sure everything came in properly
movie_table.head()

item_id              title
0        1   Toy Story (1995)
1        2   GoldenEye (1995)
2        3  Four Rooms (1995)
3        4  Get Shorty (1995)
4        5     Copycat (1995)

In [12]:
#Let's change the column headers to match up with our already established data frame and to make them slightly more expressive
column_names_2 = ['movie_id', 'movie_title']

movie_table.columns = column_names_2

#Check to make sure everything worked properly
movie_table.head()

movie_id        movie_title
0         1   Toy Story (1995)
1         2   GoldenEye (1995)
2         3  Four Rooms (1995)
3         4  Get Shorty (1995)
4         5     Copycat (1995)

In [13]:
#Now we can merge our dataframes together into a single dataframe with completely information
ratings_df_complete = ratings_df.merge(movie_table, how='left', on='movie_id')

#Check to make sure everything worked
ratings_df_complete.head(15)

user_id  movie_id  rating  timestamp  \
0         0       172       5  881250949   
1         0       133       1  881250949   
2       196       242       3  881250949   
3       186       302       3  891717742   
4        22       377       1  878887116   
5       244        51       2  880606923   
6       166       346       1  886397596   
7       298       474       4  884182806   
8       115       265       2  881171488   
9       253       465       5  891628467   
10      305       451       3  886324817   
11        6        86       3  883603013   
12       62       257       2  879372434   
13      286      1014       5  879781125   
14      200       222       5  876042340   

                                          movie_title  
0                     Empire Strikes Back, The (1980)  
1                           Gone with the Wind (1939)  
2                                        Kolya (1996)  
3                            L.A. Confidential (1997)  
4                                 Heavyweights (1994)  
5                          Legends of the Fall (1994)  
6                                 Jackie Brown (1997)  
7   Dr. Strangelove or: How I Learned to Stop Worr...  
8                    Hunt for Red October, The (1990)  
9                             Jungle Book, The (1994)  
10                                      Grease (1978)  
11                     Remains of the Day, The (1993)  
12                                Men in Black (1997)  
13      Romy and Michele's High School Reunion (1997)  
14                    Star Trek: First Contact (1996)

In [14]:
#Let's check that if a movie_id is the same, you get the same title
ratings_check = ratings_df_complete.loc[ratings_df_complete.movie_id == 302]
ratings_check.head()

user_id  movie_id  rating  timestamp               movie_title
3         186       302       3  891717742  L.A. Confidential (1997)
367       191       302       4  891560253  L.A. Confidential (1997)
704        49       302       4  888065432  L.A. Confidential (1997)
808        54       302       4  880928519  L.A. Confidential (1997)
1105       62       302       3  879371909  L.A. Confidential (1997)

Alright! Based on table and our checks, it looks we should be good to go on to EDA The last step would be to export our table so we can use it later

In [15]:
#Export as a CSV so we can begin using it
ratings_df_complete.to_csv('Ratings_Table_COMPLETE.csv')

That should complete this portion of the project! Next, we'll move onto EDA which may need to be modified a little given that we are working on a recommendation system!

Note on outliers: For a data set like this, where outliers need to be evaluted on either a row-based metric (i.e., did the user rate a bunch of movies but a specific movie rated by that users is substantially different than the other movies the user rated) or a column-based metric (i.e., a movie has gotten an in ordinate amount of ratings and a few of the ratings fall outside the range of the many ratings), it's going to be tough to really parse out whether an "outlier" is of value or something else. This is because the features we are evaluating are going to be explored in the context of "collaborative filting" where each user has preferences and it is perfectly valid for them to have "outlier" observations. During the rest of the process, we may come across items that need to be considered, but for now, it's tough to deduce if anything is genuinely an outlier.